compute-timing-stats

```diff
@@ -0,0 +1,205 @@
+
+      js_data += "[#{js_build_time}, #{js_step_durations.join(', ')}],\n"
+    end
+    js_data += "]\n"
+
+    title = "Circle CI step durations, container #{options[:container_num]}"
+
+    # Write out plot
+    # https://developers.google.com/chart/interactive/docs/gallery/areachart
+    FileUtils.mkdir_p OUTPUT_DIRECTORY
+    path = "#{OUTPUT_DIRECTORY}/report.html"
+    File.open(path, 'w') do |file|
+      file.write <<-FILE
+<html>
+  <head>
+    <script type="text/javascript" src="https://www.gstatic.com/charts/loader.js"></script>
+    <script type="text/javascript">
+      google.charts.load('current', {'packages':['corechart']});
+      google.charts.setOnLoadCallback(drawChart);
+
+      function drawChart() {
+        var data = google.visualization.arrayToDataTable(#{js_data});
+
+        var options = {
+          title: '#{title}',
+          hAxis: {title: 'Build date',  titleTextStyle: {color: '#333'}},
+          vAxis: {title: 'Step time (seconds)', minValue: 0},
+          isStacked: true
+        };
+
+        var chart = new google.visualization.AreaChart(document.getElementById('chart_div'));
+        chart.draw(data, options);
+      }
+    </script>
+  </head>
+  <body>
+    <div id="chart_div" style="width: 100%; height: 900px;"></div>
+  </body>
+</html>
+      FILE
+    end
+    Launchy.open(path)
+  else
+    # Track overall build times
+    build_durations = []
+
+    # We're going to store info here
+    # Map of step key to a list of run times
+    step_durations = {}
+
+    @project.get_builds(build_range, true).each do |build|
+      next if build['outcome'] != 'success'
+      next if build['steps'].nil? || build['build_time_millis'].nil?
+      next unless options[:branch_name].nil? || build['branch'] == options[:branch_name]
+
+      build_durations.push(build['build_time_millis'])
+
+      build['steps'].each do |step|
+        next unless options[:step_pattern].nil? || options[:step_pattern] =~ step['name']
+        step['actions'].each do |action|
+          step_key = format("%9d %s", action['index'], step['name'])
+          step_durations[step_key] ||= []
+          step_durations[step_key].push(action['run_time_millis'])
+        end
+      end
+    end
+
+    # Write out averages
+    puts "Limiting to branch #{options[:branch_name]}" unless options[:branch_name].nil?
+    puts "Average build time for #{build_durations.size} successful builds in range #{build_range}"
+    puts ms_to_readable_duration(build_durations.inject(:+).to_f / build_durations.size)
+
+    puts "\nAverage build times for steps"
+    puts format("%8s %9s %s", 'Duration', 'Container', 'Name')
+    step_durations.
+      map{|key, durations| [key, durations.inject(:+).to_f / durations.size]}.
+      sort{|a, b| b[1] <=> a[1]}.
+      each do |step|
+        puts format("%s %s", ms_to_readable_duration(step[1]), step[0])
+      end
+  end
+end
+
+MS_PER_HOUR = 3_600_000
```

Comment:
Surprised there aren't already helpers in the ruby/gem world to do these conversions...